crates/providers/src/heart.rs

```diff
@@ -16,8 +19,150 @@ use tokio::{
+        self.set_confirmations(confirmations);
+        self
+    }
+
+    /// Returns the timeout.
+    pub fn timeout(&self) -> Option<Duration> {
+        self.inner.timeout()
+    }
+
+    /// Sets the timeout.
+    pub fn set_timeout(&mut self, timeout: Option<Duration>) {
+        self.inner.set_timeout(timeout);
+    }
+
+    /// Sets the timeout.
+    pub fn with_timeout(mut self, timeout: Option<Duration>) -> Self {
+        self.set_timeout(timeout);
+        self
+    }
+
+    /// Registers the watching configuration with the provider.
+    ///
+    /// This does not wait for the transaction to be confirmed, but returns a [`PendingTransaction`]
+    /// that can be awaited at a later moment.
+    ///
+    /// See:
+    /// - [`watch`](Self::watch) for watching the transaction without fetching the receipt.
+    /// - [`get_receipt`](Self::get_receipt) for fetching the receipt after the transaction has been
+    ///   confirmed.
+    pub async fn register(self) -> TransportResult<PendingTransaction> {
+        self.provider.watch_pending_transaction(self.inner).await
+    }
+
+    /// Waits for the transaction to confirm with the given number of confirmations.
+    ///
+    /// See:
+    /// - [`register`](Self::register): for registering the transaction without waiting for it to be
+    ///   confirmed.
+    /// - [`get_receipt`](Self::get_receipt) for fetching the receipt after the transaction has been
+    ///   confirmed.
+    pub async fn watch(self) -> TransportResult<B256> {
+        self.register().await?.await
+    }
+
+    /// Waits for the transaction to confirm with the given number of confirmations, and
+    /// then fetches its receipt.
+    ///
+    /// See:
+    /// - [`register`](Self::register): for registering the transaction without waiting for it to be
+    ///   confirmed.
+    /// - [`watch`](Self::watch) for watching the transaction without fetching the receipt.
+    pub async fn get_receipt(self) -> TransportResult<Option<N::ReceiptResponse>> {
+        let pending_tx = self.provider.watch_pending_transaction(self.inner).await?;
+        let hash = pending_tx.await?;
+        self.provider.get_transaction_receipt(hash).await
+    }
+}
+
+impl<N, T, P: Clone> PendingTransactionConfig<N, T, &P> {
+    /// Clones the provider and returns a new pending transaction configuration with the cloned
+    /// provider.
+    pub fn with_cloned_provider(self) -> PendingTransactionConfig<N, T, P> {
+        PendingTransactionConfig {
+            inner: self.inner,
+            provider: self.provider.clone(),
+            _phantom: PhantomData,
+        }
+    }
+}
+
+/// A configuration object for watching for transaction confirmation.
+///
+/// This object is not directly usable, but can be used to create a [`PendingTransactionConfig`]
+/// to watch for a transaction.
+#[must_use = "this type does nothing unless you call `with_provider`"]
+#[derive(Debug)]
+pub struct PendingTransactionConfigInner {
```

Comment:
`send_transaction*` APIS currently return the Inner struct because we cannot use `&Self` as the `P: Provider` generic in the return type of a `Provider` method due to weird constraints with either object safety or `auto_impl` (or both).

This means that there's an extra `with_provider` step:
```rust
provider
.send_transaction(tx)
.await?
.with_provider(&provider) // here
.wait() // or one of the other methods in `PendingTransactionConfig`
.await?;
```

Reply:
what's the issue with auto_impl? otherwise we should prob consider implementing provider for &P manually, this API is not ideal

Reply:
Discussion here https://discord.com/channels/900503503162724452/1216893060747886673